dev/NumberBox/NumberBoxParser.cpp

```diff
@@ -0,0 +1,301 @@
+                    operatorStack.pop();
+                }
+                else
+                {
+                    break;
+                }
+            }
+            operatorStack.push(token);
+        }
+        else if (token.Type == MathTokenType::Parenthesis)
+        {
+            if (token.Char == L'(')
+            {
+                operatorStack.push(token);
+            }
+            else
+            {
+                while (!operatorStack.empty() && operatorStack.top().Char != L'(')
+                {
+                    // Pop operators onto output until we reach a left paren
+                    postfixTokens.push_back(operatorStack.top());
+                    operatorStack.pop();
+                }
+
+                if (operatorStack.empty())
+                {
+                    // Broken parenthesis
+                    error = true;
+                    break;
+                }
+
+                // Pop left paren and discard
+                operatorStack.pop();
+            }
+        }
+    }
+
+    // Pop all remaining operators.
+    while (!operatorStack.empty())
+    {
+        if (operatorStack.top().Type == MathTokenType::Parenthesis)
+        {
+            // Broken parenthesis
+            error = true;
+            break;
+        }
+
+        postfixTokens.push_back(operatorStack.top());
+        operatorStack.pop();
+    }
+
+    if (error)
+    {
+        return {};
+    }
+    return postfixTokens;
+}
+
+winrt::IReference<double> NumberBoxParser::ComputePostfixExpression(std::vector<MathToken> tokens)
+{
+    winrt::IReference<double> value = nullptr;
+    std::stack<double> stack;
+    bool error = false;
+
+    for (auto const token : tokens)
+    {
+        if (error)
+        {
+            break;
+        }
+
+        if (token.Type == MathTokenType::Operator)
+        {
+            // There has to be at least two values on the stack to apply
+            if (stack.size() < 2)
+            {
+                error = true;
+                break;
+            }
+
+            const auto op1 = stack.top();
+            stack.pop();
+
+            const auto op2 = stack.top();
+            stack.pop();
+
+            double result;
+
+            switch (token.Char)
+            {
+                case L'-':
+                    result = op2 - op1;
+                    break;
+
+                case L'+':
+                    result = op1 + op2;
+                    break;
+
+                case L'*':
+                    result = op1 * op2;
+                    break;
+
+                case L'/':
+                    if (op1 == 0)
+                    {
+                        error = true;
+                        value = NAN;
+                    }
+                    else
+                    {
+                        result = op2 / op1;
+                    }
+                    break;
+
+                case L'^':
+                    result = std::pow(op2, op1);
```

Comment:
>result = std::pow(op2, op1); [](start = 20, length = 28)

check for 0^0 as it is undefined

Reply:
That is an interesting case, but apparently it's defined as 1. TIL. I guess I'll add it to the test cases because it's a weird one.

Reply:
Hmm, my math major alter ego loves this one.

 0^0 is termed an indeterminant form that is to say that the normal calculus tools do not have a method for assigning a value to the expression.  An expression of the form a~b, where a and b are real numbers and ~ is any operation on the reals, is determinate if for all f(x) and g(x), such that the limit as x approaches 0 of f(x) = a and the limit as x approaches 0 of g(x) = b, the limit as x approaches 0 of f(x)~g(x) is well defined.

There are choices of f and g such that limit as x approaches 0 of f(x)^g(x) is not well defined, therefor 0^0 is indeterminant.

That being said other branches of math have strong arguements that 0^0 should be defined as 1, it makes many theorems behave much nicer.

finally... the standard library has decided that 0^0=1 so i guess we should to.  boring... :unamused: 

---
In reply to: [345449758](https://github.com/microsoft/microsoft-ui-xaml/pull/1602#discussion_r345449758) [](ancestors = 345449758)

Reply:
My brother was also a math major, I'm going to have to mention this to him! But yeah, calculator and the major search engines all agree it's 1.

Reply:
Yes, accept 0^0=1 into your hearts! Search your feelings, and also cardinal arithmetic and power series -- you know it to be true!

Reply:
For some purposes it is easier to just define 0^0 to be 1, however saying that 0^0=1 is not completely correct. @StephenLPeters provided a nice scenario where this is not correct. In this case setting 0^0 to 1 should be fine, even though it is not entirely correct if we look at the maths behind it.